radixdlt/src/main/java/com/radixdlt/consensus/liveness/ProposalGenerator.java

```diff
@@ -0,0 +1,37 @@
+package com.radixdlt.consensus.liveness;
+
+import com.google.common.collect.Sets;
+import com.radixdlt.common.Atom;
+import com.radixdlt.consensus.QuorumCertificate;
+import com.radixdlt.consensus.Round;
+import com.radixdlt.consensus.Vertex;
+import com.radixdlt.consensus.VertexStore;
+import com.radixdlt.mempool.Mempool;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Logic for generating new proposals
+ */
+public final class ProposalGenerator {
+	private final Mempool mempool;
+	private final VertexStore vertexStore;
+
+	@com.google.inject.Inject
+	public ProposalGenerator(
+		VertexStore vertexStore,
+		Mempool mempool
+	) {
+		this.vertexStore = Objects.requireNonNull(vertexStore);
+		this.mempool = Objects.requireNonNull(mempool);
+	}
+
+	// TODO: check that next proposal works with current vertexStore state
+	public Vertex generateProposal(Round round) {
+		QuorumCertificate highestQC = vertexStore.getHighestQC();
+
+		List<Atom> atoms = mempool.getAtoms(1, Sets.newHashSet());
```

Comment:
I feel like I have complained about this a lot, but never achieved a satisfactory resolution, nor received a satisfactory response.

The second argument to `mempool.getAtoms(...)` *cannot be a fresh set every time*, or bad things will happen.  In particular if it does not include all of the atom IDs for atoms in the pipeline, the duplicates will start to be proposed once multiple nodes are present in the system, as the mempool will distribute the atoms to all nodes immediately.

Reply:
Right, will add a fix in this PR.